### ISSUE TITLE
Subscriber of HttpClientResponse.getContent() may be unsubscribed automatically before any content is delivered

### ISSUE THREAD

**Issue by allenxwang** [ReactiveX/RxNetty#215]
I have a Ribbon test that tries to connect to a few bad servers to expect connection exception before finally hitting a good server. It then does a transformation:

``` java
Observable<Person> getPersonObservable(Observable<HttpClientResponse<ByteBuf>> response) {
        return response.flatMap(new Func1<HttpClientResponse<ByteBuf>, Observable<ByteBuf>>() {
            @Override
            public Observable<ByteBuf> call(HttpClientResponse<ByteBuf> t1) {
                System.err.println("Get response: " + t1.getStatus().code());
                return t1.getContent();
            }
        }).map(new Func1<ByteBuf, Person>() {
            @Override
            public Person call(ByteBuf t1) {
                System.err.println("Get ByteBuf");
                try {
                    return getPerson(t1); // deserialize
                } catch (IOException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        });
```

The test prints out 

```
Get response: 200
```

but the map() operation is not executed causing test to hang when it is waiting for the content as a blocking Observable.

After debugging, I found that the Observer of content (in this case the map operation) is automatically unsubscribed before any content is delivered as shown in the stack trace:

```
Daemon Thread [rx-selector-1-5] (Suspended (breakpoint at line 247 in UnicastContentSubject$OnSubscribeAction$1))   
    UnicastContentSubject$OnSubscribeAction$1.call() line: 247  
    Subscriptions$ActionSubscription.unsubscribe() line: 65 
    SubscriptionList.unsubscribeFromAll(Collection<Subscription>) line: 98  
    SubscriptionList.unsubscribe() line: 84 
    OperatorMerge$InnerSubscriber<T>(Subscriber<T>).unsubscribe() line: 62  
    CompositeSubscription.unsubscribeFromAll(Collection<Subscription>) line: 135    
    CompositeSubscription.unsubscribe() line: 125   
    SubscriptionList.unsubscribeFromAll(Collection<Subscription>) line: 98  
    SubscriptionList.unsubscribe() line: 84 
    SafeSubscriber<T>(Subscriber<T>).unsubscribe() line: 62 
    SafeSubscriber<T>.onCompleted() line: 92    
    LoadBalancerExecutor$RetryNextServerOperator$1.onCompleted() line: 152  
    LoadBalancerExecutor$RetrySameServerOperator$1.onCompleted() line: 257  
    SafeSubscriber<T>.onCompleted() line: 83    
    OperatorFinally$1.onCompleted() line: 60    
    OperatorTake$1.onNext(T) line: 66   
    SafeSubscriber<T>.onNext(T) line: 111   
    OperatorDoOnEach$1.onNext(T) line: 66   
    OperatorDoOnEach$1.onNext(T) line: 66   
    SubjectSubscriptionManager$SubjectObserver<T>.onNext(T) line: 224   
    PublishSubject<T>.onNext(T) line: 101   
    ObservableAdapter.channelRead(ChannelHandlerContext, Object) line: 40   
    DefaultChannelHandlerContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 333 
    DefaultChannelHandlerContext(AbstractChannelHandlerContext).fireChannelRead(Object) line: 319   
    ConnectionLifecycleHandler<I,O>(ChannelInboundHandlerAdapter).channelRead(ChannelHandlerContext, Object) line: 86   
    DefaultChannelHandlerContext(AbstractChannelHandlerContext).invokeChannelRead(Object) line: 333 
    DefaultChannelHandlerContext(AbstractChannelHandlerContext).fireChannelRead(Object) line: 319   
    ClientRequestResponseConverter(ChannelInboundHandlerAdapter).channelRead(ChannelHandlerContext, Object) line: 86    
    ClientRequestResponseConverter.channelRead(ChannelHandlerContext, Object) line: 120 
```

So onCompleted() of main Observable (from submit() call) is called from `OperatorTake$1.onNext(T)`, causing unsubscribing of Observer of the HttpClientResponse, as well as Observer of content. This "take 1" operation is from HttpClientImpl:

``` java
protected Observable<HttpClientResponse<O>> submit(final HttpClientRequest<I> request,
                                                       final Observable<ObservableConnection<HttpClientResponse<O>, HttpClientRequest<I>>> connectionObservable,
                                                       final ClientConfig config) {
        // ...
        return toReturn.take(1).finallyDo(new Action0() {
            @Override
            public void call() {
                eventsSubject.onEvent(HttpClientMetricsEvent.REQUEST_PROCESSING_COMPLETE,
                                      Clock.onEndMillis(startTimeMillis));
            }
        });
    }
```

It seems to me that take(1) is unnecessary here which causes the automatic onCompleted() of main Observable. It also contradicts with the code in `ClientRequestResponseConverter` where onCompleted() of main Observable is only called after all content is received.

``` java
if (LastHttpContent.class.isAssignableFrom(recievedMsgClass)) {
                eventsSubject.onEvent(HttpClientMetricsEvent.RESPONSE_RECEIVE_COMPLETE,
                                      Clock.onEndMillis(responseReceiveStartTimeMillis));
                subjectToUse.onCompleted();
                connInputObsrvr.onCompleted(); // complete the main Observable after content is delivered
            }
```

Once `take(1)` is removed, the Ribbon test passes.

**Comment by allenxwang** [comment]
After looking into Ribbon's LoadBalalancerExecutor and changing the a usage of subscribe() to unsafeSubscribe() in this line:

https://github.com/Netflix/ribbon/blob/master/ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/LoadBalancerExecutor.java#L182

the test passes without any change in RxNetty. So at this point I suspect that this is caused by incorrect usage of Rx API in Ribbon. The javadoc of unsafeSubscribe() says "Use this only for implementing an Operator that requires nested subscriptions." which is exactly what LoadBalalancerExecutor is trying to do.

**Comment by NiteshKant** [comment]
@allenxwang If your test is expecting a chunked/streaming response then this may be caused by #225 
Also see the [release notes for 0.3.12](https://github.com/ReactiveX/RxNetty/releases/tag/0.3.12) for the details about the change in behavior of `HttpClient.submit()` which is the reason for `take(1)` in `HttpClient`

**Comment by NiteshKant** [comment]
@allenxwang I will move this to 0.3.14 since I do not have reproduction steps to make sure it is fixed in 0.3.13 (as part of #225). If you can verify that 0.3.13 fixes this issue, I will close it.

**Comment by allenxwang** [comment]
Confirmed that Ribbon test passed using RxNetty 0.3.13 and with the Ribbon changes of https://github.com/Netflix/ribbon/pull/160.
